crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1692,12 +1719,18 @@ impl AuthorityPerEpochStore {
         &self,
         cert: &VerifiedExecutableTransaction,
         commit_round: Round,
-        dkg_closed: bool,
+        dkg_failed: bool,
+        generating_randomness: bool,
         previously_deferred_tx_digests: &HashMap<TransactionDigest, DeferralKey>,
         shared_object_congestion_tracker: &SharedObjectCongestionTracker,
     ) -> Option<(DeferralKey, DeferralReason)> {
-        // Defer transaction if it uses randomness but DKG has not yet closed.
-        if !dkg_closed && self.randomness_state_enabled() && cert.is_randomness_reader() {
+        // Defer transaction if it uses randomness but we aren't generating any this round.
+        // Don't defer if DKG has permanently failed; in that case we need to ignore.
+        if !dkg_failed
```

Comment:
Now reading this part closely, I think when `dkg_failed && self.randomness_state_enabled() && cert.is_randomness_reader()` is true here, we probably want to return `None` instead of checking for congestion, since the transaction will be ignored.

Other than this, deferral related logic LGTM.

Reply:
Thanks for pointing this out. I figured optimizing that case doesn't realy matter too much because it will have meant that DKG failed for the epoch and that's a huge bug that will file all sorts of alerts if it ever happens. Is the congestion checking stuff particularly expensive? If not I'd sooner keep the code simpler by not writing an extra optimization for a scenario we hope will never happen.

Reply:
I see, if DKG fail is rare, it's ok then. Maybe just keep in mind that if a transaction should be ignored, it may be unexpectedly deferred due to congestion.